src/c++/libos/lwip/lwip_queue.cc

```diff
@@ -1399,35 +1481,42 @@ dmtr::lwip_queue::parse_packet(struct sockaddr_in &src,
     dst.sin_family = AF_INET;
 
     // segment count
-    sga.sga_numsegs = ntohl(*reinterpret_cast<uint32_t *>(p));
-    p += sizeof(uint32_t);
+    uint32_t numsegs_buf;
+    auto * const numsegs = pktmbuf_struct_read(pkt, offset, numsegs_buf);
+    sga.sga_numsegs = ntohl(*numsegs);
+    offset += sizeof(uint32_t);
+
 
 #if DMTR_DEBUG
     printf("recv: sga_numsegs: %d\n", sga.sga_numsegs);
 #endif
 
     if (sga.sga_numsegs != 0xdeadbeef) {
+        // Allocate enough space for all sga segments
+        size_t sga_total_size = ntohs(ip_hdr->total_length) - sizeof(*ip_hdr) - sizeof(*udp_hdr) - sizeof(uint32_t);
+        DMTR_OK(dmtr_malloc(&sga.sga_buf, sga_total_size));
+
+        // Read into the sga buffer
+        const void *read_buf = rte_pktmbuf_read(pkt, offset, sga_total_size, sga.sga_buf);
+        // If read_buf is not pointing to sga_buf, it's because no copy was necessary
+        // (to avoid freeing problems later, copy into sga_buf anyway)
+        if (read_buf != sga.sga_buf) {
+            rte_memcpy(sga.sga_buf, read_buf, sga_total_size);
+        }
+
+        char *offset_buf = static_cast<char*>(sga.sga_buf);
         for (size_t i = 0; i < sga.sga_numsegs; ++i) {
-            // segment length
-            auto seg_len = ntohl(*reinterpret_cast<uint32_t *>(p));
+            uint32_t seg_len = ntohl(*(uint32_t*)offset_buf);
```

Comment:
Isn't it a case where we want static_cast?

Reply:
you mean instead of the c-style cast?
Since we're casting from a `void*`, either static_cast or reinterpret_cast should work. I think they compile to the same thing.
You're right though - using c-style cast is not recommended. I did it because I was lazy and think that static_cast looks ugly, but I just found this quote from Stroustrup:
```
This form of explicit type conversion is inherently unsafe and ugly. Consequently, the notation used, static_cast, was designed to be ugly.
```

Reply:
Oops, actually since `offset_buff` is a `char *`, not a `void *` (because we want to do arithmetic with it), we do have to use reinterpret_cast